### PR TITLE
ns-api: add interfaces and devices APIs

### PR DESCRIPTION
**Pull request by andre8244**
Add `ns.devices` API

See: https://github.com/NethServer/nethsecurity-ui/pull/104

### CODE REVIEW

File: packages/ns-api/files/ns.devices [hunk 1]
```diff
@@ -0,0 +1,997 @@
...
+               # discard unconfigured bond devices (they would appear after committing bond creation)
+               not is_bond(physical_dev),
+               physical_devices))
+
+    # add data from config
+
+    merged_devices = []
+
+    for physical_dev in filtered_physical_devices:
+        device_from_config_found = next((device_from_config for device_from_config in devices_from_config if physical_dev.get(
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:57]: This is almost unreadable: can we explain it with a comment or expand the code?

File: packages/ns-api/files/ns.devices [hunk 2]
```diff
@@ -0,0 +1,997 @@
...
+                'tx_packets': stat.get('stats64').get('tx').get('packets'),
+            }
+
+    # retrieve devices speed
+
+    for device in devices_to_show:
+        try:
+            res = subprocess.run(
+                ['/bin/cat', f'/sys/class/net/{device.get("name")}/speed'], check=True, text=True, capture_output=True)
+            speed = int(res.stdout.strip())
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:117]: This is slow due to the context switch: it's better to create a a function which reads the file using python without the subprocess call

File: packages/ns-api/files/ns.devices [hunk 3]
```diff
@@ -0,0 +1,997 @@
...
+            device['speed'] = speed
+        except subprocess.CalledProcessError:
+            # ignore read errors
+            pass
+
+    # sorted zones and devices
+
+    zones = [
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:125]: Why the zones are hard-coded?

File: packages/ns-api/files/ns.devices [hunk 4]
```diff
@@ -0,0 +1,997 @@
...
+
+        if not iface_found:
+            unassigned_zone['devices'].append(get_name(device))
+        else:
+            zone_found = get_firewall_zone(iface_found['.name'], fw_zones)
+
+            if not zone_found:
+                unassigned_zone['devices'].append(get_name(device))
+            else:
+                zone_obj = next((z for z in zones if z.get(
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:162]: Oh boy, another unreadable one :D

File: packages/ns-api/files/ns.devices [hunk 5]
```diff
@@ -0,0 +1,997 @@
...
+    p = subprocess.run(["/sbin/ip", "-j", "address"],
+                       check=True, text=True, capture_output=True)
+    devices = json.loads(p.stdout)
+
+    physical_devices = []
+
+    for device in devices:
+        # discard lo and ifb-dns device
+        if device.get(
+                'ifname') in ['ifb-dns', 'lo']:
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:218]: Unreadable:
```suggestion
        if device.get('ifname') in ['ifb-dns', 'lo']:
```

This is also not correct, you should skip all devices starting with `ifb-` name

File: packages/ns-api/files/ns.devices [hunk 6]
```diff
@@ -0,0 +1,997 @@
...
+                values['gateway'] = ip4_gateway
+
+                # ip6 gateway
+                if ip6_enabled:
+                    values['ip6gw'] = ip6_gateway
+                else:
+                    values['ip6gw'] = ''
+            else:
+                values['gateway'] = ''
+                values['ip6gw'] = ''
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:463]: Not sure about this: why gateway is not supported on other zones?

File: packages/ns-api/files/ns.devices [hunk 7]
```diff
@@ -0,0 +1,997 @@
...
+
+    # disable "force link" on red interfaces
+    if zone == 'wan':
+        values['force_link'] = '0'
+
+    for key, value in values.items():
+        uci.set('network', interface_name, key, value)
+
+
+def add_interface_to_firewall_zone(interface_name, zone, fw_zones, uci):
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:503]: Can we reuse functions from python3-nethsec lib? Like this one: https://nethserver.github.io/python3-nethsec/nethsec.firewall.html#add_interface_to_zone

File: packages/ns-api/files/ns.devices [hunk 8]
```diff
@@ -0,0 +1,997 @@
...
+    if not fw_zone.get('network'):
+        fw_zone['network'] = ()
+
+    # add the new interface to zone interfaces
+    fw_zone['network'] += (interface_name,)
+
+    uci.set('firewall', fw_zone['.name'], 'network', list(fw_zone['network']))
+
+
+def set_firewall_zone(interface_name, zone, interface_to_edit, uci):
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:519]: See above comment

File: packages/ns-api/files/ns.devices [hunk 9]
```diff
@@ -0,0 +1,997 @@
...
+def get_alias_interface_from_device_name(device_name, ifaces_from_config):
+    iface = get_interface(device_name, ifaces_from_config)
+
+    if not iface:
+        return None
+
+    return get_alias_interface_from_iface_name(iface['.name'], ifaces_from_config)
+
+
+def remove_interface_from_zone(iface_name, ifaces_from_config, uci):
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:663]: See https://nethserver.github.io/python3-nethsec/nethsec.firewall.html#remove_interface_from_zone

File: packages/ns-api/files/ns.devices [hunk 10]
```diff
@@ -0,0 +1,997 @@
...
+
+    if not parent_iface_name:
+        raise utils.ValidationError('parent_iface_name', 'required')
+
+    # if editing alias, ensure parent interface exists
+    if not is_creating:
+        ifaces_from_config = get_all_by_type_as_list(
+            uci, 'network', 'interface')
+        parent_iface_found = next(
+            (iface for iface in ifaces_from_config if iface.get('.name') == parent_iface_name), None)
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:792]: :scream_cat:

**Review by gsanchietti — COMMENTED** [review]
Also please remember to:
- add the new files to the Makefile
- update doc of README

File: packages/ns-api/files/ns.devices (hunk 3 above)
Comment (andre8244) [reply packages/ns-api/files/ns.devices:125]: This is just to define the sorting of the zones on the UI, other zones are appended below these with no specific order. This part will probably change soon, but we'll do it in another PR

File: packages/ns-api/files/ns.devices (hunk 6 above)
Comment (andre8244) [reply packages/ns-api/files/ns.devices:463]: Gateway is not displayed on the UI for non-wan zones, so it cannot be set or edited

File: packages/ns-api/README.md [hunk 11]
```diff
@@ -4135,3 +4135,234 @@ Invalid configuration response:
...
+```
+
+### configure-device
+
+Create or configure an unconfigured device, or edit its configuration. What happens exactly in the configuration process depends on the input parameters. Anyway, as a rule of thumb the configuration of a device executes one or more of the following steps:
+- creation of a network interface
+- association of the device to the newly created interface
+- configuration of some attributes of the network interface
+- association of the network interface to a firewall zone
+
```
Comment (gsanchietti) [inline packages/ns-api/README.md:4216]: Missing invocation example

File: packages/ns-api/README.md [hunk 12]
```diff
@@ -4135,3 +4135,234 @@ Invalid configuration response:
...
+```
+
+### unconfigure-device
+
+Remove a device to an unconfigured state by deleting the associated network interface and other data created during the configuration process.
+
+Required parameters:
+
+- `iface_name`: name of the interface associated to the device to unconfigure
+
```
Comment (gsanchietti) [inline packages/ns-api/README.md:4262]: Missing invocation example

File: packages/ns-api/README.md [hunk 13]
```diff
@@ -4135,3 +4135,234 @@ Invalid configuration response:
...
+```json
+{
+   "message": "success"
+}
+```
+
+### create-alias-interface
+
+Create an alias interface in order to associate multiple IPv4/IPv6 addresses to a network interface.
+
```
Comment (gsanchietti) [inline packages/ns-api/README.md:4274]: Missing invocation example

File: packages/ns-api/files/ns.devices (hunk 3 above)
Comment (gsanchietti) [reply packages/ns-api/files/ns.devices:132]: Ok, let's remove dmz.

File: packages/ns-api/files/ns.devices [hunk 14]
```diff
@@ -0,0 +1,1035 @@
...
+            'devices': []
+        },
+        {
+            'name': 'guests',
+            'devices': []
+        },
+        {
+            'name': 'dmz',
+            'devices': []
+        },
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:148]:
```suggestion
```

File: packages/ns-api/files/ns.devices [hunk 15]
```diff
@@ -0,0 +1,1035 @@
...
+
+def get_firewall_zone(iface_name, fw_zones):
+    zone_found = None
+
+    for zone in fw_zones:
+        if zone.get('network') and iface_name in zone.get('network'):
+            zone_found = zone
+            break
+
+    return zone_found
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:201]: Do not change it, but this is could be a more concise implementation:
```suggestion
    for zone in fw_zones:
        if zone.get('network') and iface_name in zone.get('network'):
            return zone
    return None
```

File: packages/ns-api/files/ns.devices [hunk 16]
```diff
@@ -0,0 +1,1035 @@
...
+
+def is_bridge(device):
+    if device.get('type') == 'bridge' or re.match(r"br[0-9]+", device.get('name')):
+        return True
+    else:
+        return False
+
+
+def should_omit_device_name(device_name):
+    return device_name == 'lo' or device_name.startswith('ifb-')
```
Comment (gsanchietti) [inline packages/ns-api/files/ns.devices:225]: Probably we should also skip all tun devices:
```
```suggestion
    return device_name == 'lo' or device_name.startswith('ifb-') or device_name.startswith('tun')
```
```

**Pull request merged**